src/recursive_shared_mutex.h

```diff
@@ -0,0 +1,101 @@
+#ifndef _RECURSIVE_SHARED_MUTEX_H
+#define _RECURSIVE_SHARED_MUTEX_H
+
+#include <cassert>
+#include <chrono>
+#include <condition_variable>
+#include <exception>
+#include <map>
+#include <mutex>
+#include <system_error>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+
+/*
+ * This mutex has two levels of access, shared and exclusive. Multiple threads can own this mutex in shared mode but
+ * only
+ * one can own it in exclusive mode.
+ * - A thread is considered to have ownership when it successfully calls either lock or try_lock.
+ * - A thread may recusively call lock for ownership and must call a matching number of unlock calls to end ownership.
+ * - A thread MAY call for shared ownership if it already has exclusive ownership. This should just add an additional
+ * lock on
+ * top of write counter and not actually lock. If a thread has exclusive ownership and checks for shared ownership this
+ * should return true.
+ * - A thread MAY obtain exclusive ownership if no threads excluding itself has shared ownership. (this might need to
+ * check
+ * for another write lock already
+ * queued up so we dont jump the line)
+ */
+
+
+static const std::thread::id NON_THREAD_ID = std::thread::id();
+
+class recursive_shared_mutex
+{
+private:
+    // Only locked when accessing counters, ids, or waiting on condition variables.
+    std::mutex _mutex;
+
+    // the read_gate is locked (blocked) when threads have write ownership
+    std::condition_variable _read_gate;
+
+    // the write_gate is locked (blocked) when threads have read ownership
+    std::condition_variable _write_gate;
+
+    // promotion_gates
+    std::condition_variable _promotion_read_gate;
+    std::condition_variable _promotion_write_gate;
+
+    // holds a list of owner ids that have shared ownership and the number of times they locked it
+    std::map<std::thread::id, uint64_t> _read_owner_ids;
```

Comment:
Let's imagine that all the threads used the same count variable.  This means that we couldn't detect (at the moment of the error) the situation where a thread erroneously unlocked too many times since it might not decrement the count below 0 if other threads also have the lock.  

But if we assume a correct program (in terms of matched lock/unlocks), then I think that the semantics of this object would be the same.  And our RAII (scope based) lock objects are a compile-time technique that ensures that threads match their locks and unlocks.

So depending on the performance required it might make sense to do that rather than have a map.  But the map implementation will always be valuable in a "debug" build.  

So let's keep it this way now and think about higher performing choices if needed.

Reply:
The std::map is needed for promotions. Without the map we can not efficiently know how many of the shared locks the promoting thread. 
It needs to know how many of the shared locks belong to the thread that is promoting so it can move them to the auto locks. this leaves the remaining map with all other threads that need to release shared ownership before exclusive ownership can be obtained via promotion.

Reply:
good point